pkg/services/ngalert/schedule/schedule.go

```diff
@@ -374,7 +374,7 @@ func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key ngmodels.AlertR
 		notify(states)
 	}
 
-	evaluate := func(ctx context.Context, f fingerprint, attempt int64, e *evaluation, span trace.Span) {
+	evaluate := func(ctx context.Context, f fingerprint, attempt int64, e *evaluation, span trace.Span, noRetry bool) error {
```

Comment:
Evaluate() is only called from one place. Passing in a retry flag here feels like a boundary leak.

Can this be structured such that `evaluate()` just dumbly returns an error if the resultset had an error, and then the retry logic above calls `HasNonRetryableErrors()` and determines if a retry should happen? That would also mean the span/logs gets lifted out too.

Reply:
Seconding this. Since `errors.Is()` works on joined errors, we can rewrite `HasNonRetryableErrors()` and check from the call site of `evaluate()`

```go
func HasNonRetryableErrors(err error) bool {
	return errors.Is(err, &invalidEvalResultFormatError{})
}
```

Assuming we don't want to do something with that internal count later.

Reply:
> Can this be structured such that evaluate() just dumbly returns an error if the resultset had an error, and then the retry logic above calls HasNonRetryableErrors() and determines if a retry should happen? That would also mean the span/logs gets lifted out too.

I thought about that and @rwwiv pointed out my exact thinking as well (splitting `evaluate` into two sections), but I fear this might be a much larger diff and I _really_ want to avoid that - I'd gladly pay the price and embrace the tech debt if it means I get a chance at less work as I try to backport this.

Let me know if you feel strongly about this and I'll take a stab at it, otherwise, I'd rather leave it as is if you all think the logic is acceptable.